src/CoreFoundation/CFProxySupport.cs

```diff
@@ -579,62 +581,328 @@ public static CFProxySettings GetSystemProxySettings ()
 			dict.DangerousRelease ();
 			return new CFProxySettings (dict);
 		}
-		
-#if notyet
-		// FIXME: These require bindings for CFRunLoopSource and CFStreamClientContext
-		
-		public delegate void CFProxyAutoConfigurationResultCallback (IntPtr client, NSArray proxyList, NSError error);
-		
+
+		delegate void CFProxyAutoConfigurationResultCallbackInternal (IntPtr client, IntPtr proxyList, IntPtr error);
+		public delegate void CFProxyAutoConfigurationResultCallback (NSObject client, CFProxy[] proxyList, NSError error);
+		// helper delegate to reuse code
+		delegate IntPtr CreatePACCFRunLoopSource (CFProxyAutoConfigurationResultCallbackInternal cb, ref CFStreamClientContext context);
+
+		static CFProxy[] ParseProxies (IntPtr proxyList)
+		{
+			CFProxy[] proxies = null;
+			if (proxyList != IntPtr.Zero) {
+				// it was retained in the cbs.
+				using (var array = new CFArray (proxyList, true)) {
```

Comment:
You call `ParseProxies` from two places: one is a property that can call `ParseProxies` multiple times for the same pointer (without retaining it first). This will end up crashing.

Seems like it would be simpler to pass `owns: false`, and then in the other place that calls `ParseProxies`, stop retaining the pointer first.

Reply:
I did the change and with just the Url method, the Retain deadlocks. The code paths are the same and jus the native API is changed.. I wonder what could happen. With the current implementation everything works.